feat: add strings.Replacer structure on native libraries

Villaquiranm opened the pull request:
<!-- please provide a detailed description of the changes made in this pull request. -->
prerequisite of #2802 

This pull request ports the files:
- replace.go
- replace_test.go

from the Golang standard library.

I added some tags on the code with the hope it will help to review the code and to launch discussion if neccessary.
I could after remove these changes
```go
// Custom code: XXX_Some_Explanation
( code not present on the original go file)
.
.
.
// End of custom code
```

 

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

Villaquiranm commented on gnovm/stdlibs/strings/replace.gno:
```diff
@@ -0,0 +1,622 @@
+// It is safe for concurrent use by multiple goroutines.
+type Replacer struct {
+	// Custom code: initialized bool was before sync.Once on golang package
+	initialized bool // guards buildOnce method
```
I could yes I think is a good idea, I just wanted to make the less possible changes to the original file

Villaquiranm replied:
done on 5b456eccf1e57ea700207ed7a45c935abfb57082
@petar-dambovaliev

thehowl commented on gnovm/stdlibs/strings/replace.gno:
```diff
@@ -0,0 +1,620 @@
+	return
+}
+
+// Custom code: these test functions are normally placed on export_test.go file but it seems like we do not recognize
+// methods declared on test files as part of the structure. For example:
+// Replacer.printTrie() was not recognized so I added here directly
```
wdym? what was the problem?

Villaquiranm replied:
If I deplace
```go
func (r *Replacer) PrintTrie() string {
	r.buildOnce()
	gen := r.r.(*genericReplacer)
	return gen.printNode(&gen.root, 0)
}

func (r *genericReplacer) printNode(t *trieNode, depth int) (s string) {
	if t.priority > 0 {
		s += "+"
	} else {
		s += "-"
	}
	s += "\n"

	if t.prefix != "" {
		s += Repeat(".", depth) + t.prefix
		s += r.printNode(t.next, depth+len(t.prefix))
	} else if t.table != nil {
		for b, m := range r.mapping {
			if int(m) != r.tableSize && t.table[m] != nil {
				s += Repeat(".", depth) + string([]byte{byte(b)})
				s += r.printNode(t.table[m], depth+1)
			}
		}
	}
	return
}
```
To the **export_test.go** file where it is originally on Go native lib. The method is not recognized by the test and I got the following error:
`.: test pkg: panic: strings_test/replace_test.gno:406:10: missing field PrintTrie in *strings.Replacer:`

Villaquiranm replied:
maybe there is a bug I could investigate, at least a mismatch with Go behavior

thehowl replied:
I think I get it.

When we do the import on "strings" in the testing context here, from the package `strings_test`, we don't load what is declared in the test files. (This is different from Go's behaviour, please open up an issue)

Can you put the test that uses `PrintTrie` (`TestGenericTrieBuilding` in a separate test file, together with this `PrintTrie` method, declared in `package strings`?

This way we can use PrintTrie without actually making it a publicly available method (it shouldn't be.)

Villaquiranm replied:
@thehowl fixed on b31ccc2f38674b254dd0927ae0cc70b42faed90f
Also New issue opened #2888

thehowl commented:
> I added some tags on the code with the hope it will help to review the code and to launch discussion if neccessary.
> I could after remove these changes

more than these tags, which are still good, i'm looking for tooling which can help us compare go stdlibs and gno stdlibs

there was a pr by an external contributor which was stopped at one point: https://github.com/gnolang/gno/pull/1425

if you want, you can continue his work, minding the unaddressed comments i made - i'd highly appreciate it ;)

Villaquiranm commented:
> > I added some tags on the code with the hope it will help to review the code and to launch discussion if neccessary.
> > I could after remove these changes
> 
> more than these tags, which are still good, i'm looking for tooling which can help us compare go stdlibs and gno stdlibs
> 
> there was a pr by an external contributor which was stopped at one point: #1425
> 
> if you want, you can continue his work, minding the unaddressed comments i made - i'd highly appreciate it ;)

Sure I would like to take a look @thehowl 👍
